### PR TITLE
[Fix] Add GLM-5.2 MuonSplit and AdamW-only gradient clipping

### PR DESCRIPTION
**Pull request by JT-Ushio**
## Summary

- Apply Muon independently to GLM-5.2 MLA head/component blocks, including per-block LR scaling ([GLM-5](https://arxiv.org/html/2602.15763v2#S2.SS1.SSS0.Px4), [reference implementation](https://github.com/PaddlePaddle/PaddleFormers/blob/327a4340e6c4f6fd1e3b525fb891993d23d6785a/paddleformers/transformers/glm_moe_dsa/modeling.py#L139-L252)).
- Clip only auxiliary AdamW gradients while retaining the full gradient norm for safety checks. This is motivated by Muon's scale-invariant orthogonalization, `Ortho(cG) = Ortho(G)` ([rationale](https://kellerjordan.github.io/posts/muon/#proving-that-ns-iteration-orthogonalizes-the-update)).

## Tests

- Full pre-commit suite for `xtuner/v1`, including mypy
- GPU unit tests pending maintainer approval

### CODE REVIEW

**Comment by JT-Ushio** [comment]
Rebased onto the latest `main` and fixed the mypy errors. The full pre-commit suite now passes locally. The updated fork workflows are awaiting maintainer approval before lint and unit tests can run.

**Comment by jayhenry** [comment]
当前 AdamW-only clipping 的数值行为合理，但 [`TrainEngine.clip_grad_norm()`](https://github.com/InternLM/xtuner/blob/466152f954a02e5efdbd6c53fa213f98f4faed0e/xtuner/v1/engine/train_engine.py#L257-L292) 同时承担了两层职责：高层流程需要完成“reduce gradients → 计算总范数 → 裁剪目标子集 → 上报总范数”，实现中却直接判断 `algorithm == "muon"`，并内联了子集范数、foreach 和 device fallback。这样使训练引擎知道具体 optimizer 分类，也打断了公开方法的线性业务叙事。

建议把“哪些参数需要裁剪”留在拥有 optimizer 分类知识的 `MuonConfig`，`TrainEngine` 只消费通用的 `clip_grad` policy；再将具有完整规则的分布式裁剪下沉到 `_clip_gradients()`。这个调整不改变当前训练语义，只收窄模块间的知识边界。

### 最小改进伪代码

`MuonConfig.build()` 负责把具体 optimizer 分类翻译成通用裁剪策略：

```python
param_groups = []
if muon_params_regular:
    param_groups.append({"params": muon_params_regular, "clip_grad": False})
if muon_params_moe_fused_w1w3:
    param_groups.append(
        {
            "params": muon_params_moe_fused_w1w3,
            "num_experts": 2 * num_experts,
            "clip_grad": False,
        }
    )
if muon_params_moe_other:
    param_groups.append(
        {
            "params": muon_params_moe_other,
            "num_experts": num_experts,
            "clip_grad": False,
        }
    )
param_groups.append(
    {"params": adamw_params, "algorithm": "adamw", "clip_grad": True}
)
```

`TrainEngine.clip_grad_norm()` 只保留同一抽象层级的线性流程，不再认识 Muon/AdamW：

```python
def clip_grad_norm(self, do_clip=True, dtype=torch.float32):
    ProberList.before_clip_grad_norm(self.model)
    self.model.scale_and_reduce_grad()

    all_grads = [
        parameter.grad
        for _, parameter in self.model.trainable_parameters()
        if parameter.grad is not None
    ]
    total_grad_norm = cal_grad_norm(all_grads, dtype=dtype)[0]

    if do_clip:
        clip_grads = [
            parameter.grad
            for group in self.optimizer.param_groups
            if group.get("clip_grad", True)
            for parameter in group["params"]
            if parameter.grad is not None
        ]
        self._clip_gradients(
            clip_grads,
            max_norm=self.optim_cfg.max_grad_norm,
            dtype=dtype,
        )

    ProberList.after_clip_grad_norm(self.model, total_grad_norm)
    return total_grad_norm
```

`_clip_gradients()` 隐藏“子集全局范数 + DeviceMesh/placements 分组 + foreach/device 缩放”的完整规则：

```python
def _clip_gradients(self, grads, max_norm, dtype):
    if not grads:
        return

    subset_norm, grouped_grads = cal_grad_norm(grads, dtype=dtype)
    clip_coef = torch.clamp(
        max_norm / (subset_norm + 1e-6),
        max=1.0,
    )

    for device_grads in grouped_grads.values():
        device = device_grads[0].device
        device_coef = clip_coef.to(device)
        if _device_has_foreach_support(device):
            torch._foreach_mul_(device_grads, device_coef)
        else:
            for grad in device_grads:
                grad.mul_(device_coef)
```

这里有意保留两个不同的 norm：`total_grad_norm` 继续供 `step_optimizer()` 做 NaN/Inf 和阈值跳步检查；`subset_norm` 只负责计算选定参数子集的裁剪系数。最终 `clip_grad_norm()` 保持“准备梯度 → 计算观测值 → 执行裁剪策略 → 返回观测值”的线性流程。

**Comment by JT-Ushio** [comment]
@jayhenry 已按建议在 `6e0cd9a` 完成重构：

- `MuonConfig` 将 optimizer 分类转换为通用 `clip_grad` policy；
- `TrainEngine` 只消费该 policy，并将子集 norm、foreach 和 device fallback 下沉到 `_clip_gradients()`；
- `total_grad_norm` 继续用于日志和跳步判断，AdamW 子集 norm 仅用于裁剪；普通 AdamW 复用已计算的总 norm，避免重复 all-reduce；
- `clip_grad` 作为运行时 policy 不写入 Muon checkpoint，加载时保留当前配置，兼容旧 checkpoint。

回归测试覆盖了引擎不依赖 `algorithm`、Muon 不裁剪/AdamW-only 裁剪，以及 checkpoint policy。完整 pre-commit（含 Linux 模式 mypy，359 个源码文件）已通过；GPU tests 仍待 fork workflow approval。

**Review by jayhenry — APPROVED** [review]
